### PR TITLE
🐙 octavia-cli: list connectors

### PR DESCRIPTION
**Pull request by alafanechere**
## What
We want to display a list of existing sources and destinations connectors that can then be created with (yet to be implemented) `octavia create` commands.

## How
* Declare two new subcommand groups: `octavia` > `list` > `connectors`
* Implement `sources` and `destinations` commands in the `connectors` group to list existing sources and destinations in an Airbyte instance. Make use of a `Definitions` abstract class to 
* Expose the created commands : `octavia list connectors sources` + `octavia list connectors destinations`
## Recommended reading order
1. `octavia-cli/octavia_cli/entrypoint.py` > Addition of `list` subcommand to `octavia`command group
2. `octavia-cli/octavia_cli/list/commands.py` > Declaration of the `list` and `connectors` command groups + `sources` and `destinations` commands
3. `octavia-cli/octavia_cli/list/definitions.py` Implementation of a `Definitions` abstract class to retrieve and display sources and destinations lists.
4. `octavia-cli/unit_tests/test_list/*` write unit tests for the new code.
5. `octavia-cli/unit_tests/test_entrypoint.py` update entrypoint test.

## 🚨 User Impact 🚨
`octavia --airbyte-url="https://demo.airbyte.io" list connectors sources` and `octavia --airbyte-url="https://demo.airbyte.io" list connectors destination` are now available commands and display the following output.

Screen Shot 2022-01-17 at 10 30 43

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@alafanechere these new commands are listing the current version of the containers in the user instance OR the latest version in Docker hub?

File: octavia-cli/octavia_cli/list/definitions.py
Comment (marcosmarxm) [inline octavia-cli/octavia_cli/list/definitions.py:51]: Maybe a base class to control the CLI specification where users can configure too.

File: octavia-cli/octavia_cli/list/definitions.py
Comment (marcosmarxm) [inline octavia-cli/octavia_cli/list/definitions.py:118]: Just an idea to discuss: but looks like the source/destination will probably have a similar method all along the way
```suggestion

    def __init__(self, api_client):
       super().__init__(DestinationType.SOURCE, api_client)
       self.list_latest = api_client.list_latest_source_definitions
      
      
    # move this to Definition class.
    @property
    def latest_definitions(self) -> List[List[str]]:
        api_response = self.list_latest(self.api_instance, **self.LIST_LATEST_DEFINITIONS_KWARGS)
        return self._parse_response(api_response)
```

If any function need something different btw source/destination we can overwrite and describe the reason with comments.

**Review by marcosmarxm — COMMENTED** [review]
some initial comments! It's amazing :D

File: octavia-cli/octavia_cli/list/commands.py
Comment (marcosmarxm) [inline octavia-cli/octavia_cli/list/commands.py:48]: can this be replace by a `__main__` method?

File: octavia-cli/octavia_cli/list/definitions.py
Comment (marcosmarxm) [inline octavia-cli/octavia_cli/list/definitions.py:18]:
```suggestion
class ConnectorDefinitions(abc.ABC):
```
Maybe use a more specific naming compared to Definition? Just a suggestion

File: octavia-cli/octavia_cli/list/definitions.py
Comment (marcosmarxm) [inline octavia-cli/octavia_cli/list/definitions.py:67]:
```suggestion
    def _camelcased_to_uppercased_spaced(camelcased: str) -> str:
```

Comment (marcosmarxm) [reply octavia-cli/octavia_cli/list/definitions.py:67]: As you already mention this can be moved to a util/misc file to be used in other places.

File: octavia-cli/octavia_cli/list/definitions.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/list/definitions.py:51]: Yes, you're right. I did not want to make premature optimizations and think that this refactoring need and the right place to put these functions will become more evident during the implementation of other commands.

File: octavia-cli/octavia_cli/list/definitions.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/list/definitions.py]: I think I went in the direction you suggested in this commit: d515b94de77351b977b14b19776118018d2790e8

File: octavia-cli/octavia_cli/list/definitions.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/list/definitions.py]: I renamed it to `ConnectorsDefinitions`: d515b94de77351b977b14b19776118018d2790e8

File: octavia-cli/octavia_cli/list/commands.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/list/commands.py:48]: No, because I want this to run when it's imported by `entrypoint.py`

File: octavia-cli/octavia_cli/list/definitions.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/list/definitions.py]: Yes, it'll be done in the following PR.

**Comment by alafanechere** [comment]
I opened the next PR (as draft) that implements a lot of the `TODO` mentioned in this one and make heavy use of inheritance to have a shared logic between listing commands: https://github.com/airbytehq/airbyte/pull/9642

**Comment by alafanechere** [comment]
> @alafanechere these new commands are listing the current version of the containers in the user instance OR the latest version in Docker hub?

We are using this [endpoint](https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/source_definitions/list_latest) with I think merges GitHub's `source_definitions.yaml` file and the custom connector.

File: octavia-cli/octavia_cli/list/commands.py
Comment (lmossman) [inline octavia-cli/octavia_cli/list/commands.py:29]: @alafanechere it looks like these commands are currently implemented to list source/destination connector **definitions**, i.e. list all of the connector docker images that the user can use to actually create a "Source" or a "Destination". 

I think the naming here might be a bit misleading, as the command is `octavia list connectors sources` rather than something like `octavia list connectors source_definitions`. It may also be useful to have a true `list sources` command that returns the Sources that have actually been created in the workspace, i.e. the result of this API endpoint: https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/sources/list. And similarly for Destinations.

What do you think?

File: octavia-cli/octavia_cli/entrypoint.py
Comment (lmossman) [inline octavia-cli/octavia_cli/entrypoint.py:38]: Is there a reason why you defined this method to add commands to `octavia`, and then later call it, rather than just adding the commands directly in the `def octavia()` method?

Comment (lmossman) [reply octavia-cli/octavia_cli/entrypoint.py:38]: There could very well be some technical reason why you can't do the latter that I am unaware of!

File: octavia-cli/octavia_cli/list/commands.py
Comment (lmossman) [reply octavia-cli/octavia_cli/list/commands.py:29]: Thank you for the explanation! That makes a lot more sense to me now - I think the naming of those commands looks good, and I agree that the word `connector` is more meaningful to the user than `definition`.

File: octavia-cli/octavia_cli/entrypoint.py
Comment (alafanechere) [reply octavia-cli/octavia_cli/entrypoint.py:38]: This a bit specific to the `Click` package we are using.
Here `octavia` is a command **group**, and `list` is a command group too.
It's the way I found to add a command group to another.

Usually to add a command (and not command group) to a group the decorator `@<command-group-name>.command()` works well, e.g to declare a `octavia create` command: 
```python 
@octavia.command(help="Generate a YAML configuration file to manage a resource.")
def create() -> None:
    ....
````

I could try if this work in `octavia-cli/octavia_cli/list/commands.py`:
```
@octavia.group
def _list():
  pass
```

But I like the fact of referencing everything that's added to the main `octavia` command to `octavia-cli/octavia_cli/entrypoint.py`, which will make it more readable to understand the flow.

Comment (lmossman) [reply octavia-cli/octavia_cli/entrypoint.py:38]: I see, I think that makes sense! I have no familiarity with the `Click` package, so I will defer to your judgement on the best approach to take here

File: octavia-cli/octavia_cli/list/connectors_definitions.py
Comment (lmossman) [inline octavia-cli/octavia_cli/list/connectors_definitions.py:52]: I agree with these comments that these methods should go in a formatting module, they seem very useful to reuse in other commands 👍

File: octavia-cli/octavia_cli/list/connectors_definitions.py
Comment (lmossman) [inline octavia-cli/octavia_cli/list/connectors_definitions.py:121]: I like how you did this inheritance setup - you were able to avoid repeating 99% of the shared logic between source and destination definitions. Nice job!

File: octavia-cli/unit_tests/test_list/test_connectors_definitions.py
Comment (lmossman) [inline octavia-cli/unit_tests/test_list/test_connectors_definitions.py:105]: As noted above, these tests should also be moved to a formatting module when you move the underlying methods

**Review by lmossman — APPROVED** [review]
Looks good to me! Though it looks like there are some build errors that should be fixed before merging

**Review by marcosmarxm — APPROVED** [review]

**Comment by alafanechere** [comment]
Thank you @marcosmarxm and @lmossman for the review and your valuable feedback!
The build error is due to some formatting inconsistencies between `isort` in pre-commit hook and `isortFormat` task ran by `./gradlew format` (line length parameter is not set to 140 in pre-commit). I discarded pre-commit changes for `isort` in this PR, and will check if we can align formatting task and pre-commit hook parameters in another PR, because it might affect other parts of the project.

**Pull request merged**
